File: 542.01-matrix.py
# @leet start
from collections import deque
from typing import List


class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        queue = deque()
        nrows = len(mat)
        ncols = len(mat[0])
        for r, row in enumerate(mat):
            for c, ele in enumerate(row):
                if ele == 0:
                    queue.append((r, c))

        # ms-bfs
        visited = set()
        dist = -1

        while queue:
            dist += 1
            for _ in range(len(queue)):
                r, c = queue.popleft()
                if (r, c) in visited:
                    continue

                visited.add((r, c))
                mat[r][c] = dist

                for dr, dc in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                    rn = r + dr
                    cn = c + dc

                    if 0 <= rn < nrows and 0 <= cn < ncols:
                        queue.append((rn, cn))

        return mat


# @leet end


m0 = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
e0 = [[0, 0, 0], [0, 1, 0], [0, 0, 0]]
m1 = [[0, 0, 0], [0, 1, 0], [1, 1, 1]]
e1 = [[0, 0, 0], [0, 1, 0], [1, 2, 1]]
x = Solution().updateMatrix(m1)
print(x)
print(x == e1)

File: 542.01-matrix.java
// @leet start
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        // TC: O(m*n)
        // SC: O(1)
        int nrows = mat.length;
        int ncols = mat[0].length;
        int inf = 1_000_000_000;

        for (int i = 0; i < nrows; i++) {
            for (int j = 0; j < ncols; j++) {
                if (mat[i][j] == 0) {
                    continue;
                }
                int top = inf;
                int left = inf;
                if (i - 1 >= 0) {
                    top = mat[i - 1][j];
                }
                if (j - 1 >= 0) {
                    left = mat[i][j-1];
                }
                mat[i][j] = 1 + Math.min(top, left);
            }
        }

        for (int i = nrows - 1; i >= 0; i--) {
            for (int j = ncols - 1; j >= 0; j--) {
                if (mat[i][j] == 0) {
                    continue;
                }
                int bottom = inf;
                int right = inf;
                if (i + 1 < nrows) {
                    bottom = mat[i + 1][j];
                }
                if (j + 1 < ncols) {
                    right = mat[i][j + 1];
                }
                // Two mins because mat[i][j] could be lower from the above loop
                mat[i][j] = Math.min(mat[i][j], 1 + Math.min(bottom, right));
            }
        }

        return mat;

    }
}
// @leet end
